File: js/problem41.js
/**
 * @param {number[]} nums
 * @return {number}
 */


const checkOnePresent = function(nums) {
    for (let i = 0; i < nums.length; i++) {
        if (nums[i] === 1) {
            return true;
        }
    }
    return false;
};

var firstMissingPositive = function(nums) {
    if (nums.length === 0) {
        return 1;
    } else if (nums.length === 1) {
        if (nums[0] === 1) {
            return 2;
        }
        return 1;
    }
    
    if (!checkOnePresent(nums)) {
        return 1;
    }
    
    for (let i = 0; i < nums.length; i++) {
        if (nums[i] < 1) {
            nums[i] = 1;
        }
    }
    
    let nPresent = false;
    
    for (let i = 0; i < nums.length; i++) {
        let temp = Math.abs(nums[i]);
        if (temp >= nums.length) {
            if (temp === nums.length) {
                nPresent = true;
            }
            continue;
        }
        
        if (nums[temp] > 0) {
            nums[temp] *= -1; 
        }
    }
    
    for (let i = 1; i < nums.length; i++) {
        if (nums[i] > 0) {
            return i;
        }
    }
    
    if (!nPresent) {
        return nums.length;
    }
    
    return nums.length + 1;
};

File: leetcodeJava/src/main/java/Problem41.java
// 41. First Missing Positive

class Solution {
    
    private boolean containsOne(int[] nums) {
        for (int n : nums) {
            if (n == 1) {
                return true;
            }
        }
        return false;
    }
    
    public int firstMissingPositive(int[] nums) {
        if (nums.length == 1) {
            if (nums[0] != 1) {
                return 1;
            }
            return 2;
        }
        
        if (!this.containsOne(nums)) {
            return 1;
        }
        
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] < 1) {
                nums[i] = 1;
            }
        }
        
        for (int i = 0; i < nums.length; i++) {
            int current = Math.abs(nums[i]);
            if (current > nums.length || current == 1) {
                continue;
            } else if (current == nums.length) {
                if (nums[0] > 0) {
                    nums[0] *= -1;
                }
            } else {
                if (nums[current] > 0) {
                    nums[current] *= -1;
                }
            }
        }
        
        for (int i = 2; i < nums.length; i++) {
            if (nums[i] > 0) {
                return i;
            }
        }
        
        if (nums[0] > 0) {
            return nums.length;
        }
        return nums.length + 1;
    }
}